### ISSUE TITLE
Internal Server Error 500

### ISSUE THREAD

**Issue by karafecho** [NCATS-Tangerine/tranql#48]
![image](https://user-images.githubusercontent.com/26254388/71038763-960e9780-20f0-11ea-86b6-7bfc59e65e06.png)

**Comment by mmersmann** [comment]
@karafecho Can you paste in the query text that produced this issue?

**Comment by karafecho** [comment]
The query is behind the error box and described within that box.

Essentially, I'm having lots of transition issues with transitions that use to work. May be related to BioLink. In the example here, the transition issue produced an internal server error. Steve asked Yaphet and Chuck to investigate.

**Comment by YaphetKG** [comment]
We were running a query something of this nature
 ``` 
SELECT population_of_individual_organisms->disease
          FROM "/clinical/cohort/disease_to_chemical_exposure"
         WHERE icees.table = 'patient'
           AND icees.year = 2010
           AND icees.feature.EstResidentialDensity = 1 
          AND icees.maximum_p_value = 1
``` 
and this returned a value , I think icees expects icees.feature to be set

**Comment by YaphetKG** [comment]
```
SELECT population_of_individual_organisms->disease->gene
         FROM "/schema"
        WHERE icees.table = 'patient'
          AND icees.year = 2010
          AND icees.feature.EstResidentialDensity = 1 
         AND icees.maximum_p_value = 1
``` 
And also this has response coming from robokop, maybe icees.feature is a required parameter

**Comment by YaphetKG** [comment]
this is a response segment from Icees 
```
  "response": {
    "version": "2.0.0",
    "return value": {
      "code_description": "'feature'",
      "datetime": "2019-12-12/18/19 20:41:41",
      "message_code": "Error",
      "reasoner_id": "ICEES",
      "tool_version": "2.0.0"
    },
``` 
Returned when running a machine question 
```
 {
  "query_options": {
    "table": "patient",
    "year": 2010,
    "maximum_p_value": 1
  },
  "machine_question": {
    "edges": [
      {
        "id": "e1",
        "source_id": "population_of_individual_organisms",
        "target_id": "disease",
        "type": "association"
      }
    ],
    "nodes": [
      {
        "id": "population_of_individual_organisms",
        "type": "population_of_individual_organisms"
      },
      {
        "id": "disease",
        "type": "disease"
      }
    ]
  }
}
```

**Comment by stevencox** [comment]
@karafecho - we can't see the full text of the query. Could you paste the text here?

**Comment by karafecho** [comment]
I think I ran the query below:

select population_of_individual_organisms->disease
  from "/schema"
 where icees.table="patient"
 and icees.year="2010"
 and icees.cohort_feature={}
 and icees.maximum_p_value=1
   AND disease !=~ '^(SCTID.*|MONDO.*| umlscui.*)$'

**Comment by karafecho** [comment]
I ran a number of variations of the same general query. All failed. I think Yaphet may be correct that icees.feature is a required parameter.

**Comment by karafecho** [comment]
Also, I think the FROM clause has changed and needs to be specified more precisely.

**Comment by karafecho** [comment]
This query runs but does not transition to ROBOKOP:

SELECT population_of_individual_organisms->disease
         FROM "/clinical/cohort/disease_to_chemical_exposure"
        WHERE icees.table = 'patient'
          AND icees.year = 2010
          AND icees.feature.Sex2 = 'Male' 
         AND icees.maximum_p_value = 1

This query does not run:

SELECT population_of_individual_organisms->disease->chemical_substance
         FROM "/clinical/cohort/disease_to_chemical_exposure"
        WHERE icees.table = 'patient'
          AND icees.year = 2010
          AND icees.feature.Sex2 = 'Male' 
         AND icees.maximum_p_value = 1

Neither does this one:

SELECT population_of_individual_organisms->disease->chemical_exposure
         FROM "/clinical/cohort/disease_to_chemical_exposure"
        WHERE icees.table = 'patient'
          AND icees.year = 2010
          AND icees.feature.Sex2 = 'Male' 
         AND icees.maximum_p_value = 1

**Comment by karafecho** [comment]
This query appears to get to ROBOKOP via disease->gene->biological_process_or_activity, but then breaks down.

SELECT population_of_individual_organisms->disease->gene->biological_process_or_activity->gene->chemical_substance
         FROM "/clinical/cohort/disease_to_chemical_exposure"
        WHERE icees.table = 'patient'
          AND icees.year = 2010
          AND icees.feature.Sex2 = 'Male' 
         AND icees.maximum_p_value = 1

![image](https://user-images.githubusercontent.com/26254388/71125902-cf5d0b00-21b5-11ea-9bd9-c3321cf3c442.png)

**Comment by YaphetKG** [comment]
I think robokop does have gene-> biological_process_or_activity but not biological_process_or_activity -> gene,

**Comment by karafecho** [comment]
Thanks. The transition was available at one point. Is this list of [ROBOKOP operations](https://robokop.renci.org/api/operations/) up to date? I would think that it would be, given the recent NCATS request for testing and documentation, but I'm not certain that it is, or perhaps I'm misreading a few of the operations. Shall I post this question to the ROBOKOP team?

Also, I'd like to better understand the structure of the FROM clause in TranQL, as this appears to have changed.

@YaphetKG : Perhaps we can touch base to discuss the issues I've posted? That way, I could run a few queries and show you the full set of error messages.

**Comment by stevencox** [comment]
When typing a query, pressing Ctrl-Space after an arrow shows the valid BioLink-Model concepts.

So the valid transitions from gene are:

   ![image](https://user-images.githubusercontent.com/306971/71218785-d7d64400-2290-11ea-9986-605eeed7a9a2.png)
 
And valid ones from biological_process_or_activity:

   ![image](https://user-images.githubusercontent.com/306971/71218864-2552b100-2291-11ea-8a82-bdc3b0179064.png)

It also works for predicates: 

   ![image](https://user-images.githubusercontent.com/306971/71218895-44e9d980-2291-11ea-89b3-3e4fbda8f5ff.png)

**Comment by karafecho** [comment]
This TranqL example query does not run:

![image](https://user-images.githubusercontent.com/26254388/71223663-b5e5bd00-22a2-11ea-893e-1a55297ad1bc.png)

**Comment by karafecho** [comment]
![image](https://user-images.githubusercontent.com/26254388/71223862-0eb55580-22a3-11ea-9726-cb3a4bf57fa5.png)

**Comment by karafecho** [comment]
Thanks for the tip, @stevenco re "When typing a query, pressing Ctrl-Space after an arrow shows the valid BioLink-Model concepts". But, I'm not sure how a user would figure that out. Plus, the BioLink data model is evolving, which makes R&D challenging.

**Comment by YaphetKG** [comment]
@karafecho, Yeah sure, I'll be glad to help mean while I'll try to find what is going on with the example query

**Comment by karafecho** [comment]
@stevencox : I met with Yaphet earlier today. Issues were resolved within about five minutes. 

One issue was related to the FROM clause. My original query and thinking was correct in that it should be set to "/schema" for execution of query planning. 

A second issue was related to an incorrect constraint in the example query. **This still needs to get corrected in the Workflow 5 v3 example query documentation.**

A third issue related to an ICEES constraint and a specific feature variable.

So, all is fine (for now).
